File: pr-2632/sphinx/targets/python/oqc.py
import cudaq

# You only have to set the target once! No need to redefine it
# for every execution call on your kernel.
# To use different targets in the same file, you must update
# it via another call to `cudaq.set_target()`

# To use the OQC target you will need to set the following environment variables
# OQC_URL
# OQC_EMAIL
# OQC_PASSWORD
# To setup an account, contact oqc_qcaas_support@oxfordquantumcircuits.com

cudaq.set_target("oqc")


# Create the kernel we'd like to execute on OQC.
@cudaq.kernel
def kernel():
    qvector = cudaq.qvector(2)
    h(qvector[0])
    x.ctrl[qvector[1], qvector[1]]
    mz(qvector)


# Option A:
# By using the asynchronous `cudaq.sample_async`, the remaining
# classical code will be executed while the job is being handled
# by OQC. This is ideal when submitting via a queue over
# the cloud.
async_results = cudaq.sample_async(kernel)
# ... more classical code to run ...

# We can either retrieve the results later in the program with
# ```
# async_counts = async_results.get()
# ```
# or we can also write the job reference (`async_results`) to
# a file and load it later or from a different process.
file = open("future.txt", "w")
file.write(str(async_results))
file.close()

# We can later read the file content and retrieve the job
# information and results.
same_file = open("future.txt", "r")
retrieved_async_results = cudaq.AsyncSampleResult(str(same_file.read()))

counts = retrieved_async_results.get()
print(counts)

# Option B:
# By using the synchronous `cudaq.sample`, the execution of
# any remaining classical code in the file will occur only
# after the job has been returned from OQC.
counts = cudaq.sample(kernel)
print(counts)

File: docs/sphinx/targets/cpp/oqc.cpp
// Compile and run with:
// ```
// nvq++ --target oqc oqc.cpp -o out.x && ./out.x
// ```
// This will submit the job to the OQC platform. You can also specify
// the machine to use via the `--oqc-machine` flag:
// ```
// nvq++ --target oqc --oqc-machine lucy oqc.cpp -o out.x && ./out.x
// ```
// The default is the 8 qubit Lucy device. You can set this to be either
// `toshiko` or `lucy` via this flag.
//
// To use the OQC target you will need to set the following environment
// variables: OQC_URL OQC_EMAIL OQC_PASSWORD To setup an account, contact
// oqc_qcaas_support@oxfordquantumcircuits.com

#include <cudaq.h>
#include <fstream>

// Define a simple quantum kernel to execute on OQC backends.
struct bell_state {
  auto operator()() __qpu__ {
    cudaq::qvector q(2);
    h(q[0]);
    x<cudaq::ctrl>(q[0], q[1]);
  }
};

int main() {
  // Submit to OQC asynchronously (e.g., continue executing
  // code in the file until the job has been returned).
  auto future = cudaq::sample_async(bell_state{});
  // ... classical code to execute in the meantime ...

  // Can write the future to file:
  {
    std::ofstream out("future.json");
    out << future;
  }

  // Then come back and read it in later.
  cudaq::async_result<cudaq::sample_result> readIn;
  std::ifstream in("future.json");
  in >> readIn;

  // Get the results of the read in future.
  auto async_counts = readIn.get();
  async_counts.dump();

  // OR: Submit to OQC synchronously (e.g., wait for the job
  // result to be returned before proceeding).
  auto counts = cudaq::sample(bell_state{});
  counts.dump();
}
